sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,322 @@
+        Return the labels (1 inlier, -1 outlier) of X according to LOF score
+        and the contamination parameter.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier Factor
+            wrt to the training samples.
+
+        Returns
+        -------
+        is_inlier : array of shape (n_samples,)
+            Returns 1 for anomalies/outliers and -1 for inliers.
+        """
+
+        return self.fit(X)._predict()
+
+    def fit(self, X, y=None):
+        """Fit the model using X as training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix, BallTree, KDTree}
+            Training data. If array or matrix, shape [n_samples, n_features],
+            or [n_samples, n_samples] if metric='precomputed'.
+        """
+        if not (0. < self.contamination <= .5):
+            raise ValueError("contamination must be in (0, 0.5]")
+
+        super(LocalOutlierFactor, self).fit(X)
+
+        self._k_distance_value_fit_X_, self.neighbors_indices_fit_X_ = (
+            self.kneighbors(None))
+
+        # Compute decision_function over training samples to define threshold_:
+        self.outlier_factor_ = self._local_outlier_factor()
+
+        self.threshold_ = -scoreatpercentile(
+            self.outlier_factor_, 100. * (1. - self.contamination))
+
+        # XXX may be optimized (if X is not None) by only computing it for
+        # X_sub_samples = neighbors of neighbors of X ?
+        return self
+
+    def _predict(self, X=None):
+        """Predict the labels (1 inlier, -1 outlier) of X according to LOF.
+        If X is None, returns the same as fit_predict(X_train).
+        This method allows to generalize prediction to new observations (not
+        in the training set).
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier Factor
+            wrt to the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        is_inlier : array of shape (n_samples,)
+            Returns 1 for anomalies/outliers and -1 for inliers.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_k_distance_value_fit_X_",
+                               "neighbors_indices_fit_X_"])
+
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
+            is_inlier = np.ones(X.shape[0], dtype=int)
+            is_inlier[self.decision_function(X) <= self.threshold_] = -1
+        else:
+            is_inlier = np.ones(self._fit_X.shape[0], dtype=int)
+            is_inlier[-self.outlier_factor_ <= self.threshold_] = -1
+
+        return is_inlier
+
+    def decision_function(self, X):
```

Comment:
Maybe I miss something in the discussion but why don't call that `score` as usual ?